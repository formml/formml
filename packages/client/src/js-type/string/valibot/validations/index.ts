import bool from './bool.js'
import datetime from './datetime.js'
import decimal from './decimal.js'
import num from './num.js'

export { bool, datetime, decimal, num }
