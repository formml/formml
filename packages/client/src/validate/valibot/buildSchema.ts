import { Field } from '@formml/dsl'

export default function buildSchema(formmlSchema: Field) {
  return formmlSchema
}
