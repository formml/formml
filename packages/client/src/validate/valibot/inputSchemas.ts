import * as v from 'valibot'

export const number = () => v.special<string>((input) => !isNaN(Number(input)))
