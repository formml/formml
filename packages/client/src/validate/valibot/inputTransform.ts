import { BigNumber } from 'bignumber.js'
import dayjs from 'dayjs'
import * as v from 'valibot'

export const toNum = (schema?: v.BaseSchema) =>
  v.transform<v.StringSchema, number>(
    v.string([v.custom((i) => !isNaN(Number(i)))]),
    Number,
    schema,
  )
export const toDatetime = (schema?: v.BaseSchema) =>
  v.transform<v.StringSchema, dayjs.Dayjs>(
    v.string([v.isoTimestamp()]),
    dayjs,
    schema,
  )
export const toBool = (schema?: v.BaseSchema) =>
  v.transform<v.StringSchema, boolean>(v.string(), Boolean, schema)
export const toDecimal = (schema?: v.BaseSchema) =>
  v.transform<v.StringSchema, BigNumber>(
    v.string([v.custom((i) => !new BigNumber(i).isNaN())]),
    (i) => new BigNumber(i),
    schema,
  )
