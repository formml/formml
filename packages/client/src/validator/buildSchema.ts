import { Annotation, Field, PrimitiveType, resolveArguments } from '@formml/dsl'
import annotationsInterface from '@formml/dsl/interfaces/annotations.js'
import { BigNumber } from 'bignumber.js'
import * as v from 'valibot'

import { PrimitiveTypeMapping } from '../JsTypes.js'
import { assertNever } from '../utils/assertNever.js'
import annotationsReducer, { IAnnotationAction } from './annotationsReducer.js'

const type = (formmlSchema: Field) => {
  if (formmlSchema.type === 'text') {
    return v.string()
  }
  if (formmlSchema.type === 'num') {
    return v.number()
  }
  if (formmlSchema.type === 'bool') {
    return v.boolean()
  }
  if (formmlSchema.type === 'datetime') {
    return v.date()
  }
  if (formmlSchema.type === 'decimal') {
    return v.instance(BigNumber)
  }
  return assertNever`Unsupported field schema: ${formmlSchema}`
}

function buildAction(annotation: Annotation): IAnnotationAction {
  // TODO: get info from real referred declaration and ensure type safety
  const annotationName = annotation.call.$refText as IAnnotationAction['name']
  return {
    name: annotationName,
    options: resolveArguments(
      annotation.args,
      annotationsInterface[annotationName],
    ),
  }
}

export default function buildSchema<T extends PrimitiveType>(
  formmlSchema: Field<T>,
): v.GenericSchema<PrimitiveTypeMapping[T]>
export default function buildSchema(formmlSchema: Field) {
  const baseSchema = v.optional(type(formmlSchema))
  const actions = formmlSchema.annotations.map(buildAction)
  return actions.reduce(annotationsReducer, baseSchema)
}
