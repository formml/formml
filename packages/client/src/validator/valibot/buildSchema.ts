import { Field, PrimitiveType } from '@formml/dsl'
import { BigNumber } from 'bignumber.js'
import dayjs from 'dayjs'
import * as v from 'valibot'

import { PrimitiveTypeMapping } from '../../JsTypes.js'
import { assertNever } from '../../utils/assertNever.js'

const type = (
  formmlSchema: Field,
  options?: Partial<{
    notEmpty: boolean
  }>,
) => {
  if (formmlSchema.type === 'text') {
    if (options?.notEmpty) {
      return v.pipe(v.string(), v.nonEmpty())
    }
    return v.string()
  }
  if (formmlSchema.type === 'num') {
    return v.number()
  }
  if (formmlSchema.type === 'bool') {
    return v.boolean()
  }
  if (formmlSchema.type === 'datetime') {
    return v.custom<dayjs.Dayjs>(dayjs.isDayjs)
  }
  if (formmlSchema.type === 'decimal') {
    return v.instance(BigNumber)
  }
  return assertNever`Unsupported field schema: ${formmlSchema}`
}

const isRequired = (formmlSchema: Field) =>
  formmlSchema.annotations.some((a) => a.name === 'required')

export default function buildSchema<T extends PrimitiveType>(
  formmlSchema: Field<T>,
): v.GenericSchema<PrimitiveTypeMapping[T]>
export default function buildSchema(formmlSchema: Field) {
  if (isRequired(formmlSchema)) {
    return type(formmlSchema, { notEmpty: true })
  }
  return v.optional(type(formmlSchema))
}
