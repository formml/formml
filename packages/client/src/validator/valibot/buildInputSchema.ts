import { Field } from '@formml/dsl'

import { assertNever } from '../../utils/assertNever.js'
import buildSchema from './buildSchema.js'
import * as t from './transform/index.js'

export default function buildInputSchema(formmlSchema: Field) {
  if (formmlSchema.type === 'text') {
    const typedSchema = buildSchema(formmlSchema)
    return typedSchema
  }

  if (formmlSchema.type === 'num') {
    const typedSchema = buildSchema(formmlSchema)
    return t.toNum(typedSchema)
  }

  if (formmlSchema.type === 'bool') {
    const typedSchema = buildSchema(formmlSchema)
    return t.toBool(typedSchema)
  }

  if (formmlSchema.type === 'datetime') {
    const typedSchema = buildSchema(formmlSchema)
    return t.toDatetime(typedSchema)
  }

  if (formmlSchema.type === 'decimal') {
    const typedSchema = buildSchema(formmlSchema)
    return t.toDecimal(typedSchema)
  }

  return assertNever`Unsupported field schema: ${formmlSchema}`
}
