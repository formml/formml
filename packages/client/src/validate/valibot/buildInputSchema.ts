import { Field } from '@formml/dsl'

import { assertNever } from '../../utils/assertNever.js'
import buildSchema from './buildSchema.js'
import * as i from './inputTransform.js'

export default function buildInputSchema(formmlSchema: Field) {
  if (formmlSchema.type === 'text') {
    return buildSchema(formmlSchema)
  }

  if (formmlSchema.type === 'num') {
    return i.toNum()
  }

  if (formmlSchema.type === 'bool') {
    return i.toBool()
  }

  if (formmlSchema.type === 'datetime') {
    return i.toDatetime()
  }

  if (formmlSchema.type === 'decimal') {
    return i.toDecimal()
  }

  return assertNever`Unsupported field type: ${formmlSchema.type}`
}
