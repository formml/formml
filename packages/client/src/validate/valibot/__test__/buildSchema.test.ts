import { Field, Form } from '@formml/dsl'
import { BigNumber } from 'bignumber.js'
import dayjs from 'dayjs'
import * as v from 'valibot'

import buildSchema from '../buildSchema.js'

describe('buildSchema', () => {
  describe('integration', () => {
    describe.each(['text', 'num', 'bool', 'datetime', 'decimal'] as const)(
      'types - %s',
      (type) => {
        const fixtures = {
          bool: {
            expectedType: 'boolean',
            invalidInput: 'abc',
            validInput: true,
          },
          datetime: {
            expectedType: 'dayjs',
            invalidInput: new Date('2024-01-01T00:00:00Z'),
            validInput: dayjs('2024-01-01T00:00:00Z'),
          },
          decimal: {
            expectedType: 'BigNumber',
            invalidInput: 123.45,
            validInput: new BigNumber('123.45'),
          },
          num: {
            expectedType: 'number',
            invalidInput: 'abc',
            validInput: 123,
          },
          text: {
            expectedType: 'string',
            invalidInput: 123,
            validInput: 'abc',
          },
        }

        test(`should validate ${fixtures[type].expectedType} input if field type is ${type}`, () => {
          // Arrange
          const field: Field = {
            $container: {} as Form,
            $type: 'Field',
            annotations: [],
            name: 'field',
            type,
          }

          // Act
          const schema = buildSchema(field)
          const result = v.safeParse(schema, fixtures[type].validInput)

          // Assert
          expect(result.success).toBe(true)
        })

        test(`should invalidate non-${fixtures[type].expectedType} input if field type is ${type}`, () => {
          // Arrange
          const field: Field = {
            $container: {} as Form,
            $type: 'Field',
            annotations: [],
            name: 'field',
            type,
          }

          // Act
          const schema = buildSchema(field)
          const result = v.safeParse(schema, fixtures[type].invalidInput)

          // Assert
          expect(result.success).toBe(false)
          expect(result.issues).toMatchSnapshot()
        })
      },
    )
  })
})
