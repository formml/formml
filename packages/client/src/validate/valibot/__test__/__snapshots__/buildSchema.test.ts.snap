// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`buildSchema > integration > required > should invalidate undefined if field is required 1`] = `
[
  {
    "abortEarly": undefined,
    "abortPipeEarly": undefined,
    "context": "string",
    "expected": "string",
    "input": undefined,
    "issues": undefined,
    "lang": undefined,
    "message": "Invalid type: Expected string but received undefined",
    "path": undefined,
    "reason": "type",
    "received": "undefined",
    "skipPipe": undefined,
  },
]
`;

exports[`buildSchema > integration > required > should invalidate undefined if field is required 2`] = `
[
  {
    "abortEarly": undefined,
    "abortPipeEarly": undefined,
    "context": "number",
    "expected": "number",
    "input": undefined,
    "issues": undefined,
    "lang": undefined,
    "message": "Invalid type: Expected number but received undefined",
    "path": undefined,
    "reason": "type",
    "received": "undefined",
    "skipPipe": undefined,
  },
]
`;

exports[`buildSchema > integration > required > should invalidate undefined if field is required 3`] = `
[
  {
    "abortEarly": undefined,
    "abortPipeEarly": undefined,
    "context": "boolean",
    "expected": "boolean",
    "input": undefined,
    "issues": undefined,
    "lang": undefined,
    "message": "Invalid type: Expected boolean but received undefined",
    "path": undefined,
    "reason": "type",
    "received": "undefined",
    "skipPipe": undefined,
  },
]
`;

exports[`buildSchema > integration > required > should invalidate undefined if field is required 4`] = `
[
  {
    "abortEarly": undefined,
    "abortPipeEarly": undefined,
    "context": "special",
    "expected": "unknown",
    "input": undefined,
    "issues": undefined,
    "lang": undefined,
    "message": "Invalid type: Expected unknown but received undefined",
    "path": undefined,
    "reason": "type",
    "received": "undefined",
    "skipPipe": undefined,
  },
]
`;

exports[`buildSchema > integration > required > should invalidate undefined if field is required 5`] = `
[
  {
    "abortEarly": undefined,
    "abortPipeEarly": undefined,
    "context": "instance",
    "expected": "BigNumber",
    "input": undefined,
    "issues": undefined,
    "lang": undefined,
    "message": "Invalid type: Expected BigNumber but received undefined",
    "path": undefined,
    "reason": "type",
    "received": "undefined",
    "skipPipe": undefined,
  },
]
`;

exports[`buildSchema > integration > types - bool > should invalidate non-boolean input if field type is bool 1`] = `
[
  {
    "abortEarly": undefined,
    "abortPipeEarly": undefined,
    "context": "boolean",
    "expected": "boolean",
    "input": "abc",
    "issues": undefined,
    "lang": undefined,
    "message": "Invalid type: Expected boolean but received "abc"",
    "path": undefined,
    "reason": "type",
    "received": ""abc"",
    "skipPipe": undefined,
  },
]
`;

exports[`buildSchema > integration > types - datetime > should invalidate non-dayjs input if field type is datetime 1`] = `
[
  {
    "abortEarly": undefined,
    "abortPipeEarly": undefined,
    "context": "special",
    "expected": "unknown",
    "input": 2024-01-01T00:00:00.000Z,
    "issues": undefined,
    "lang": undefined,
    "message": "Invalid type: Expected unknown but received Date",
    "path": undefined,
    "reason": "type",
    "received": "Date",
    "skipPipe": undefined,
  },
]
`;

exports[`buildSchema > integration > types - decimal > should invalidate non-BigNumber input if field type is decimal 1`] = `
[
  {
    "abortEarly": undefined,
    "abortPipeEarly": undefined,
    "context": "instance",
    "expected": "BigNumber",
    "input": 123.45,
    "issues": undefined,
    "lang": undefined,
    "message": "Invalid type: Expected BigNumber but received 123.45",
    "path": undefined,
    "reason": "type",
    "received": "123.45",
    "skipPipe": undefined,
  },
]
`;

exports[`buildSchema > integration > types - num > should invalidate non-number input if field type is num 1`] = `
[
  {
    "abortEarly": undefined,
    "abortPipeEarly": undefined,
    "context": "number",
    "expected": "number",
    "input": "abc",
    "issues": undefined,
    "lang": undefined,
    "message": "Invalid type: Expected number but received "abc"",
    "path": undefined,
    "reason": "type",
    "received": ""abc"",
    "skipPipe": undefined,
  },
]
`;

exports[`buildSchema > integration > types - text > should invalidate non-string input if field type is text 1`] = `
[
  {
    "abortEarly": undefined,
    "abortPipeEarly": undefined,
    "context": "string",
    "expected": "string",
    "input": 123,
    "issues": undefined,
    "lang": undefined,
    "message": "Invalid type: Expected string but received 123",
    "path": undefined,
    "reason": "type",
    "received": "123",
    "skipPipe": undefined,
  },
]
`;
